File: Exceptions/Ruby/StandardError/code.rb
# 1
def print_exception(exception, explicit)
    puts "[#{explicit ? 'EXPLICIT' : 'INEXPLICIT'}] #{exception.class}: #{exception.message}"
    puts exception.backtrace.join("\n")
end

begin
    raise "Uh oh!"
rescue IndexError => e
    print_exception(e, true)
rescue NameError => e
    print_exception(e, true)
rescue RegexpError => e
    print_exception(e, true)
rescue => e
    print_exception(e, false)
end


# 2
class MissingName < StandardError
    def initialize(msg="Name is missing!")
        super
    end
end

def print_exception(exception, explicit)
    puts "[#{explicit ? 'EXPLICIT' : 'INEXPLICIT'}] #{exception.class}: #{exception.message}"
    puts exception.backtrace.join("\n")
end

begin
    raise MissingName
rescue => e
    print_exception(e, true)
end

File: Exceptions/JavaScript/InternalError/TooMuchRecursion/code.js
// 1
var countdown = function(value) {
    console.log(value);
    return (value > 0) ? countdown(value - 1) : value;
};
countdown(10);

// 2
var printError = function(error, explicit) {
    console.log(`[${explicit ? 'EXPLICIT' : 'INEXPLICIT'}] ${error.name}: ${error.message}`);
}
var countdown = function(value) {
    //console.log(value);
    try {
        if (value > 0) countdown(value - 1);
    } catch (e) {
        if (e instanceof InternalError) {
            printError(e, true);
        } else {
            printError(e, false);
        }
    }
};
countdown(10000);
